tianshou/policy/modelbase/psrl.py

```diff
@@ -0,0 +1,223 @@
+        rew_square_sum: np.ndarray,
+        rew_count: np.ndarray,
+    ) -> None:
+        """Add data into memory pool.
+
+        For rewards, we have a normal prior at first. After we observed a
+        reward for a given state-action pair, we use the mean value of our
+        observations instead of the prior mean as the posterior mean. The
+        standard deviations are in inverse proportion to the number of the
+        corresponding observations.
+
+        :param np.ndarray trans_count: the number of observations, with shape
+            (n_state, n_action, n_state).
+        :param np.ndarray rew_sum: total rewards, with shape
+            (n_state, n_action).
+        :param np.ndarray rew_square_sum: total rewards' squares, with shape
+            (n_state, n_action).
+        :param np.ndarray rew_count: the number of rewards, with shape
+            (n_state, n_action).
+        """
+        self.updated = False
+        self.trans_count += trans_count
+        sum_count = self.rew_count + rew_count
+        self.rew_mean = (self.rew_mean * self.rew_count + rew_sum) / sum_count
+        self.rew_square_sum += rew_square_sum
+        raw_std2 = self.rew_square_sum / sum_count - self.rew_mean ** 2
+        self.rew_std = np.sqrt(1 / (
+            sum_count / (raw_std2 + self.__eps) + 1 / self.rew_std_prior ** 2))
+        self.rew_count = sum_count
+
+    @staticmethod
+    def sample_from_prob(trans_count: np.ndarray) -> np.ndarray:
+        sample_prob = np.zeros_like(trans_count)
+        n_s, n_a = trans_count.shape[:2]
+        for i in range(n_s):
+            for j in range(n_a):  # numba does not support dirichlet :(
+                sample_prob[i][j] = np.random.dirichlet(trans_count[i][j])
+        return sample_prob
+
+    def sample_from_rew(self) -> np.ndarray:
+        return np.random.normal(self.rew_mean, self.rew_std)
+
+    def solve_policy(self) -> None:
+        self.updated = True
+        self.policy, self.value = self.value_iteration(
+            self.sample_from_prob(self.trans_count),
+            self.sample_from_rew(),
+            self.discount_factor,
+            self.eps,
+            self.value,
+        )
+
+    @staticmethod
+    def value_iteration(
+        trans_prob: np.ndarray,
+        rew: np.ndarray,
+        discount_factor: float,
+        eps: float,
+        value: np.ndarray,
+    ) -> np.ndarray:
+        """Value iteration solver for MDPs.
+
+        :param np.ndarray trans_prob: transition probabilities, with shape
+            (n_state, n_action, n_state).
+        :param np.ndarray rew: rewards, with shape (n_state, n_action).
+        :param float eps: for precision control.
+        :param float discount_factor: in [0, 1].
+        :param np.ndarray value: the initialize value of value array, with
+            shape (n_state, ).
+
+        :return: the optimal policy with shape (n_state, ).
+        """
+        Q = rew + discount_factor * trans_prob.dot(value)
```

Comment:
Correct me if I'm wrong (I'm not familiar with PSRL). This variable name is trans_prob, so I assume it is transition prob p(s' | s, a). But the output of ``sample_from_prob `` seems to be samples, not probability. Why is it this way?

Reply:
That's a PSRL feature: it samples an MDP each time and solve a policy.

Reply:
Then do not call it ``trans_prob``.